pkg/core/image.go

```diff
@@ -122,37 +155,69 @@ func relocateManifest(manifestPath string, mapper *imageMapper, outputPath strin
 		return err
 	}
 
+	nonCollidingFlattener := findNonCollidingFlattener(manifest)
+
+	if nonCollidingFlattener == nil {
+		return errors.New("cannot relocate manifest due to collisions in output paths")
+	}
+
 	manifestDir := filepath.Dir(manifestPath)
 
 	outputManifest := &Manifest{
 		ManifestVersion: manifest.ManifestVersion,
 	}
 
-	outputManifest.Istio, err = relocateYamls(manifest.Istio, manifestDir, mapper, outputPath)
+	outputManifest.Istio, err = relocateYamls(manifest.Istio, manifestDir, mapper, outputPath, nonCollidingFlattener)
 	if err != nil {
 		return err
 	}
-	outputManifest.Knative, err = relocateYamls(manifest.Knative, manifestDir, mapper, outputPath)
+	outputManifest.Knative, err = relocateYamls(manifest.Knative, manifestDir, mapper, outputPath, nonCollidingFlattener)
 	if err != nil {
 		return err
 	}
-	outputManifest.Namespace, err = relocateYamls(manifest.Namespace, manifestDir, mapper, outputPath)
+	outputManifest.Namespace, err = relocateYamls(manifest.Namespace, manifestDir, mapper, outputPath, nonCollidingFlattener)
 	if err != nil {
 		return err
 	}
 
-	outputManifestPath := filepath.Join(outputPath, "manifest.yaml")
+	outputManifestPath := filepath.Join(outputPath, nonCollidingFlattener("./manifest.yaml"))
 	outputManifestBytes, err := yaml.Marshal(&outputManifest)
 	if err != nil {
 		return err
 	}
 	return ioutil.WriteFile(outputManifestPath, outputManifestBytes, outputFilePermissions)
 }
 
-func relocateYamls(yamls []string, manifestDir string, mapper *imageMapper, outputPath string) ([]string, error) {
+func findNonCollidingFlattener(manifest *Manifest) uriFlattener {
+	// check that no new fields have been added to the manifest to avoid maintainability issues
+	if reflect.Indirect(reflect.ValueOf(manifest)).Type().NumField() != 4 {
+		panic("unexpected number of fields in manifest")
+	}
+
+	var nonCollidingFlattener uriFlattener = nil
+flatteners:
+	for _, f := range flatteners {
+		unflattened := make(map[string]string)
+		for _, array := range [][]string{manifest.Istio, manifest.Knative, manifest.Namespace, {"./manifest.yaml"}} {
```

Comment:
This whole function is a bit tough to digest, but I think the biggest problem is that it assumes that the input filename of the manifest is `manifest.yaml`, while it could be anything.

Or is that a potential candidate __output__ name? (not correctly reflected in the var names below).